### PR TITLE
Convert the Table components to FC/Hooks

### PR DESCRIPTION
**Pull request by colinlohner**
## Overview 
Table and all sub components converted to use hooks and memo.
 
## Screenshots
localhost_6006__path=_story_image--image(Laptop with HiDPI screen)
localhost_6006__path=_story_image--image(Laptop with HiDPI screen) (1)
localhost_6006__path=_story_image--image(Laptop with HiDPI screen) (2)

## Testing
- [x] Updated Typescript types and/or component PropTypes

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -1700,15 +1697,15 @@ interface TableEditableCellProps extends Omit<TextTableCellProps, 'placeholder'
...
+export class TableHeaderCell extends React.FC<TableHeaderCellProps> {
 }
 
 export interface TableHeadProps extends PaneProps {
   height?: number | string
   accountForScrollbar?: boolean
 }
 
-export class TableHead extends React.PureComponent<TableHeadProps> {
+export class TableHead extends React.FC<TableHeadProps> {
```
Comment (mshwery) [inline index.d.ts:1708]: If any of these are using memo or forwardRef you should use `ForwardRefComponent` custom type.

These aren't classes, so you'll want something like this:
https://github.com/segmentio/evergreen/blob/v5/index.d.ts#L1118

Comment (mshwery) [reply index.d.ts:1708]: If you really just want React.FC you dont need `class Foo extends`

File: src/table/src/EditableCell.js [hunk 2]
```diff
@@ -1,190 +1,168 @@
...
+    disabled,
+    placeholder,
+    isSelectable = true,
+    textProps = {},
+    autoFocus = false,
+    ...rest
+  } = props
+
+  let cursor = 'text'
+  let mainRef
```
Comment (mshwery) [inline src/table/src/EditableCell.js:23]: Would put this in state instead.

File: src/table/src/EditableCell.js [hunk 3]
```diff
@@ -1,190 +1,168 @@
...
-  handleFieldCancel = () => {
-    this.setState({ isEditing: false })
+  const handleFieldCancel = () => {
+    setIsEditing(false)
   }
 
-  handleClick = () => {
-    this.mainRef.focus()
+  const handleClick = () => {
+    mainRef.focus()
```
Comment (mshwery) [inline src/table/src/EditableCell.js:77]: This needs a `if (mainRef) {`

File: src/table/src/SelectMenuCell.js [hunk 4]
```diff
@@ -1,197 +1,166 @@
...
+  const onResizeHandler = () => {
+    updateOnResize()
   }
 
-  constructor(props) {
-    super(props)
-    this.onResize = debounce(this.onResize, 200)
-  }
+  const onResize = debounce(onResizeHandler, 200)
+  let mainRef
```
Comment (mshwery) [inline src/table/src/SelectMenuCell.js:32]: Same here, `useState` for any refs

**Comment by colinlohner** [comment]
Gonna split this up into separate PRs for the different sub components of Table

**Pull request closed without merging**